src/Pagination/PaginateDirective.php

```diff
@@ -164,13 +166,11 @@ protected function paginationType(): PaginationType
 
     protected function defaultCount(): ?int
     {
-        return $this->directiveArgValue('defaultCount')
-            ?? config('lighthouse.pagination.default_count');
+        return $this->directiveArgValue('defaultCount', config('lighthouse.pagination.default_count'));
```

Comment:
I must admit that I changed practically all calls to `directiveArgValue` to use `??` for defaults some time ago. Now, I recognize this was the wrong thing to do, as it disallows explicitly set `null`.

I documented the method, can you change other calls back to providing `$default`?

Reply:
I can't see any other directiveArgValue ?? usage where a null value would be preferable than the actual ?? fallback

```graphql
@guard(with: null)
@auth(guard: null)
@whereAuth(guard: null)
@cache(private: null)
@globalId(decode: null)
@orderBy(direction: null)
@like/@eq/@where/... (key:null)
@where (operator:null, clause:null)
...
```

None of them make sense to me. I mean, I don't expect particular behavior compared to not setting attribute at all.

---

BUT, as it looks like keeping null value (like here for maxCount, and defaultCount) is really uncommon, possibly we could change implementation of directiveArgValue to return fallback when null value are provided in schema. And add a new method hasDirectiveArg($name) to handle maxCount and defaultCount

```php
    return $this->hasDirectiveArg('defaultCount')
        ? $this->directiveArgValue('defaultCount')
        : config('lighthouse.pagination.default_count'));
```

And so change all others `directiveArgValue($name) ?? $default` to `directiveArgValue($name, $default)`

Reply:
> And so change all others directiveArgValue($name) ?? $default to directiveArgValue($name, $default)

Given you checked the other use cases, it might not be necessary, but I also see no downside to doing just that. I would always expect explicit `null` to be preserved over any defaults that might exist.

Reply:
> And add a new method hasDirectiveArg($name) to handle maxCount and defaultCount

`directiveHasArgument()` exists and is used in some places - sometimes it makes sense - but the `$default` argument handles the case you described just fine.

Reply:
> it might not be necessary, but I also see no downside to doing just that
Doing this may introduce some new breaking change.

Let's take example of clause argument on where directive.

```php
        $clause = $this->directiveArgValue('clause') ?? 'where';
        return $builder->{$clause}(...)
```

Let's then suppose the following weird notation in graphQL schema :

```graphql
@where(clause: null)
```

This notation is pretty weird, but it "works" today, clause is set to "where", thanks to ?? 'where'.

If we change all other directiveArgValue($name) ?? $default to directiveArgValue($name, $default), the notation above will end up with a PHP Error: Method name must be a string.

Reply:
I am fine with breaking nonsensical uses of explicit `null`. In the future `webonyx/graphql-php` will gain the ability to validate the correctness of schema directive usage statically.

Reply:
I re-read carefully [GraphQL specification #sec-Type-System.Non-Null](http://spec.graphql.org/June2018/#sec-Type-System.Non-Null). I think I was wrong with this pull request.

> Inputs (such as field arguments), are always optional by default. However a non‐null input type is required. In addition to not accepting the value null, it also does not accept omission. For the sake of simplicity nullable types are always optional and non‐null types are always required.

it is not 100% clear if an omitted attribute is conceptually equivalent or not to attribute: null. But maybe doing this differentiation is not correct. If conceptually `maxCount: null` is equivalent of no maxCount parameter at all (which fallback to settings) this PR should NOT change the current behavior.

This being said. We are speaking here about the server schema, not the exposed GraphQL server. Following strictly "by the book" the GraphQL spec for server schema might not be a requirement.

Reply:
> it is not 100% clear if an omitted attribute is conceptually equivalent or not to attribute: null